crates/qpy/src/circuit_writer.rs

```diff
@@ -0,0 +1,910 @@
+                base_gate = operation.operation.bind(py).getattr("base_op")?.clone();
+            }
+        } else {
+            match operation.view() {
+                // all-around catch for "operation" field; should be easier once we switch from python to rust
+                OperationRef::Gate(pygate) => {
+                    let gate = pygate.gate.bind(py);
+                    match getattr_or_none(gate, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                OperationRef::Instruction(pyinst) => {
+                    let inst = pyinst.instruction.bind(py);
+                    match getattr_or_none(inst, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                OperationRef::Operation(pyoperation) => {
+                    let operation = pyoperation.operation.bind(py);
+                    match getattr_or_none(operation, "definition") {
+                        None => (),
+                        Some(definition) => {
+                            has_definition = true;
+                            data = serialize(&pack_circuit(
+                                &mut definition.extract()?,
+                                Some(py.None().bind(py)),
+                                false,
+                                qpy_data.version,
+                                qpy_data.annotation_handler.annotation_factories,
+                            )?);
+                        }
+                    }
+                }
+                _ => (),
+            }
+        }
+        let num_qubits = operation.num_qubits();
+        let num_clbits = operation.num_clbits();
+        if !base_gate.is_none() {
+            let instruction = qpy_data
+                .circuit_data
+                .pack(py, &CircuitInstruction::py_new(&base_gate, None, None)?)?;
```

Comment:
We can probably avoid calling `CircuitData::pack` in this case by instatiating `PackedInstruction` diectly and make sure we just use an empty `Vec<formats::CircuitInstructionArgPack>` in `pack_instruction` instead of calling `get_packed_bit_list`. 
This change will allow using the input `CircuitData` as an immutable borrow throughout the writer code.